### ISSUE TITLE
RAM memory leaking

### ISSUE THREAD

**Issue by nosound2** [lyft/l5kit#144]
This is my training loop and my RAM utilization keeps growing during the run:

```
losses_train = []
tr_it = iter(train_dataloader)
progress_bar = tqdm(range(cfg["train_params"]["max_num_steps"]))
st = time.time()

torch.set_grad_enabled(True)

for k in progress_bar:
    
    data_cpu = next(tr_it)
    data = {v:k.to(device) for v,k in data_cpu.items()}
    for v,k in data_cpu.items():
        del k
    del data_cpu
    
    model.train()
    loss, _ = forward(data, model, device, criterion)

    optimizer.zero_grad()
    loss.backward()
    optimizer.step()
    
    loss_cpu = loss.detach().item()

    losses_train.append(loss_cpu)
    progress_bar.set_description("loss: % 3.2f loss(avg): % 3.2f"%(loss_cpu,np.mean(losses_train)))
    
    del data, loss, loss_cpu
    
print("running time", time.time() - st)
```

Running this cell again initially releases the memory, but then it is growing again. The memory is released specifically on call
`tr_it = iter(train_dataloader)`. Running on `Ubuntu 20.04`, jupyter notebook, config is the following:

```
{'key': 'scenes/train.zarr',
 'batch_size': 16,
 'shuffle': True,
 'num_workers': 16}
```

Thank you in advance!

**Comment by lucabergamini** [comment]
Hey @nosound2 ,
What's the extent of the leak in terms of GB?

I'm aware of some minor leaks due to the use of LRU_cache and multiprocessing. In that situation (when `num_workers`>0) the context is copied around together with the cache, so with 16 workers it will hit 16Gb after some time (as far as I remember we have a 1GB cache around). Not really sure that something we can fix though.

However, if you're experiencing a HUGE RAM leak then there may be something else going on I'm not aware of!

**Comment by nosound2** [comment]
Thanks a lot for the clarification, actually the "leak" is capped at around 16Gb, so it fits your explanation. I thought it continued growing, but now I run a longer test and I see that there is a limit. 

It is not a leak, I am closing the issue. Thanks!

**Comment by ryanchesler** [comment]
I have encountered something similar. With 16 workers on each epoch the memory continues to grow. I thought something was incorrect in my validation process, but turning off my validation process did not remove the leak. In my scenario I have to use the torch.multiprocessing.set_sharing_strategy('file_system'), I believe this is to do with me using a dual processor system, but I think that might be leading to the leak for me. Going to retry on a different system and see if it persists. 

I have seen my system continue to climb until all memory is consumed though after iterating across 2000 batches of size 32 for 4 epochs

**Comment by lucabergamini** [comment]
Hey @ryanchesler, what is the reason behind the change of multiprocessing strategy?

**Comment by ryanchesler** [comment]
When I use the default strategy I run into some weird runtime error with "RuntimeError: received 0 items of ancdata". That has been true of many other projects in pytorch on my system though so I have gotten accustomed to just adding that simple line at the beginning of my notebooks. 

I am testing with workers set to 0 to see if the leak persists, I assume it will go away. Hard to test because it will likely be around 16 hours before I see the same accumulation I was seeing with workers set to 16. 

I have multiple other things to test on my agenda, just reporting so others can see more details on the issue.

**Comment by lucabergamini** [comment]
Are you by any means appending data from torch (even after numpy cast) to a list? see [this](https://github.com/pytorch/pytorch/issues/41465) maybe. I'm not saying is your case (ancdata is a super generic error) but maybe...

**Comment by ryanchesler** [comment]
Hmm I am appending a loss to a list after I detach, push to cpu and then cast to numpy. I did see someone mention elsewhere to use numpy arrays, but I presumed that only really mattered for inside of the dataloader. I will try commenting out that section and see if it persists

**Comment by ryanchesler** [comment]
https://www.kaggle.com/ryches/memory-leak?scriptVersionId=42484552

I created this minimal example that seems to gain memory as it iterates through the dataloader. Is there something obviously wrong with my code here? Seems like this should be fine.

**Comment by lucabergamini** [comment]
> Hmm I am appending a loss to a list after I detach, push to cpu and then cast to numpy

nah that should be fine in my experience



> I created this minimal example that seems to gain memory as it iterates through the dataloader. Is there something obviously wrong with my code here?

What's the extent of the leak? Because we have caching in place the memory is expected to grow. 
Do you still experience it if you use:
```python
ChunkedDataset("PATH").open(cached=False)
```

**Comment by ryanchesler** [comment]
Responded on kaggle but I will respond here as well incase anyone else is looking here. I tried with caching turned off on kaggle and it seems to still leak over time. Across 30 minutes of iteration it went from 6gb to 9gb of usage and was continuing to climb. Did not allow it to continue running until crash but I assume it would have similar fate to the other trials I have done where kernel eventually dies

**Comment by lucabergamini** [comment]
thanks for taking the time to test this @ryanchesler!
I'm reopening this and I'll have a look into it this week!

**Comment by louis925** [comment]
I encountered the same memory leak issue on Windows 10, where I set the `num_workers` to 0 and `.open(cached=False)`, without actually training loading any pytorch but just iterate through the dataset loader from l5kit. 
It leads to a nonpaged memory leak (>10GB), even though the actual working set and private bytes of the `python.exe` is much smaller (only about 4GB).
This seems to come from that python use lots of handles without closing it. There are like 1M+ handles under `python.exe`, whereas regular program like Chrome typically use at most 4k handles. 
So perhaps, the memory leak is caused by the parallelization.

**Comment by lucabergamini** [comment]
I was not able to pinpoint this to any specific components sadly. Note that if this is only happening under Windows there is not a lot we can do...we don't have any Windows developer right now :(

**Comment by louis925** [comment]
I think I figured it out after debugging for like 2 days! 
It is the random line:
`os.environ["BLOSC_NOLOCK"] = "1"`
in the code.
https://github.com/lyft/l5kit/blob/0c5d1c5123ead9f5ce25bea6042f1062667a9fc0/l5kit/l5kit/dataset/select_agents.py#L26
This terrible line will modify your system variable secretly whenever you import any module of `l5kit.dataset`.
Once you remove this line, on Windows you will no longer see the exploding number of handles created by python.

**Comment by lucabergamini** [comment]
> It is the random line:
> `os.environ["BLOSC_NOLOCK"] = "1"`

This is not a random line, it's actually required by zarr when doing multiprocessing . I can remove it for Windows, but you may encounter random locks, corrupted files and other nasty issues afterwards while writing the mask file

**Comment by lucabergamini** [comment]
@louis925 take a look at #194 :)

**Comment by lucabergamini** [comment]
I've merged #194. However, my personal suggestion is to use Linux here and not Windows, life should be much more easy xD.

I'll close this for now, if everyone is still experiencing the issue please feel free to reopen and let me know
